src/interactivity/win32/Clipboard.cpp

```diff
@@ -230,46 +233,48 @@ void Clipboard::StoreSelectionToClipboard(const bool copyFormatting)
     const auto& renderSettings = gci.GetRenderSettings();
 
     const auto GetAttributeColors = [&](const auto& attr) {
-        return renderSettings.GetAttributeColors(attr);
+        const auto [fg, bg] = renderSettings.GetAttributeColors(attr);
+        const auto ul = renderSettings.GetAttributeUnderlineColor(attr);
+        return std::tuple{ fg, bg, ul };
     };
 
-    bool includeCRLF, trimTrailingWhitespace;
+    bool singleLine = false;
     if (WI_IsFlagSet(OneCoreSafeGetKeyState(VK_SHIFT), KEY_PRESSED))
     {
         // When shift is held, put everything in one line
-        includeCRLF = trimTrailingWhitespace = false;
+        singleLine = true;
     }
-    else
+
+    const auto& [selectionStart, selectionEnd] = selection.GetSelectionAnchors(true);
+
+    const auto req = TextBuffer::CopyRequest::FromConfig(buffer, selectionStart, selectionEnd, singleLine, !selection.IsLineSelection(), false);
+    text = buffer.GetPlainText(req);
+
+    if (copyFormatting)
     {
-        includeCRLF = trimTrailingWhitespace = true;
+        const auto& fontData = gci.GetActiveOutputBuffer().GetCurrentFont();
+        const auto& fontName = fontData.GetFaceName();
+        const auto fontSizePt = fontData.GetUnscaledSize().height * 72 / ServiceLocator::LocateGlobals().dpi;
+        const auto bgColor = renderSettings.GetAttributeColors({}).second;
+        const auto isIntenseBold = renderSettings.GetRenderMode(::Microsoft::Console::Render::RenderSettings::Mode::IntenseIsBold);
+
+        htmlData = buffer.GenHTML(req, fontSizePt, fontName, bgColor, isIntenseBold, GetAttributeColors);
+        rtfData = buffer.GenRTF(req, fontSizePt, fontName, bgColor, isIntenseBold, GetAttributeColors);
     }
 
-    const auto text = buffer.GetText(includeCRLF,
-                                     trimTrailingWhitespace,
-                                     selectionRects,
-                                     GetAttributeColors,
-                                     !selection.IsLineSelection());
-
-    CopyTextToSystemClipboard(text, copyFormatting);
+    CopyTextToSystemClipboard(text, std::move(htmlData), std::move(rtfData));
 }
 
 // Routine Description:
 // - Copies the text given onto the global system clipboard.
 // Arguments:
-// - rows - Rows of text data to copy
-// - fAlsoCopyFormatting - true if the color and formatting should also be copied, false otherwise
-void Clipboard::CopyTextToSystemClipboard(const TextBuffer::TextAndColor& rows, const bool fAlsoCopyFormatting)
+// - text - plain-text data
+// - htmlData - HTML copy data
+// - rtfData - RTF copy data
+void Clipboard::CopyTextToSystemClipboard(const std::wstring_view text, const std::optional<std::string> htmlData, const std::optional<std::string> rtfData) const

```

Comment:
I believe you can use `std::string_view` for `htmlData` and `rtfData` and use `!htmlData.empty()`, because setting the HTML/RTF clipboard to an empty string isn't useful. I believe the same can be done with `TextCopyData`.

Reply:
Dustin also pointed out the same thing, and we decided to leave it like that. However, I had already made all the changes and was about to push it before he made the offer to keep the std::optional as-is.

Since we all agree to this as a welcome change, I just pulled the changes back from the stash and pushed it 😄

Reply:
Ohh... and instead of using `std::string_view` which doesn't guarantee null-termination, I simply used `const std::string&` in all the places. This ensures copying the text over to Clipboard is safe.